### DISCUSSION TITLE
SecondsSince / execution delay

### DISCUSSION THREAD

**Discussion by CMDR-Evolution** [EDCD/EDDI#2456] [EDDI Standalone Q&A] [answered]
Greetings all,

Thank you to everyone who's involved in this amazing software. Honestly, Elite would feel so empty without EDDI. I always feel bad for people who don't know about this great software as they're missing out on so much. I wish Frontier would shine a spotlight on this great work of art. 

I'm sorry to ask about this as the post isn't to highlight a bug or issue, but is to ask for some help (if possible) please.

- Regarding the 'Carrier Stats' script - this is currently blank by default and I've put a custom script together. 
- The trouble is, it runs every time we view the fleet carrier window and as the in-game screens are within the main FC screen - the EDDI script is being executed too much. 
- So, I've tried two approaches to add a delay to force it not to run if it's just been executed. The following script has both my attempts shown - can you see where I'm going wrong please?

```
{if SecondsSince(F("Carrier stats timer")) / 1000 > 1:

{if SecondsSince(eddi_context_approach_time) > 1000:

{set notorious to OneOf("wanted","criminal","notorious")}
{if event.notoriousAccess:
	{set notoriousaccess to cat(", including ", notorious," commanders")}
|else:
	{set notoriousaccess to ""}
}
{if event.dockingAccess = "all":
	{set dockingaccess to cat("all", notoriousaccess)}
|elif event.dockingAccess = "squadronfriends {notoriousaccess}":
	{set dockingaccess to "squadron members and friendly commanders {notoriousaccess}"}
|elif event.dockingAccess = "friends {notoriousaccess}":
	{set dockingaccess to "friendly commanders only {notoriousaccess}"}
|else:
	{set dockingaccess to "no commanders except yourself"}
}

{if event.fuel = 1000:
	{set tritium to "100" }
|else:
	{set tritiumcalc to div(event.fuel, 1000)}
	{set tritium to mul(tritiumcalc, 100)}
}

{set bankcalc to div(event.bankReservedBalance, event.bankBalance)}
{set bank to mul(bankcalc, 100)}

{set cargocalc to div(event.freeCapacity, 25000)}
{set cargo to mul(cargocalc, 100)}


Welcome to the {OneOf("carrier","fleet carrier")}, {event.name}.

This carrier currently allows access to {dockingaccess}.

Tritium levels are {event.fuel} tonnes or {round(tritium)} percent.

{OneOf("The carrier","The fleet carrier","{event.name}")}'s {OneOf("free","available","un-used","remaining","un-allocated")} cargo 
{OneOf("capacity","space","room")} is {round(cargo)} percent, at {event.freeCapacity} tonnes, with {event.usedCapacity} tonnes {OneOf("used","un-available","already allocated")}.

The total bank balance is {Humanise(event.bankBalance)}, with a {round(bank)} percent reserve of {Humanise(event.bankReservedBalance)}, 
leaving an available balance of {Humanise(event.bankAvailableBalance)} credits.

}

{F("Carrier stats timer")}
{SetState('eddi_context_approach_time', SecondsSince(0))}

}

```

**Comment by CMDR-Evolution** [comment]
To try to explain my ways of thinking of the two approaches which are both not working, so far:

### First attempt was:
`{if SecondsSince(F("Carrier stats timer")) / 1000 > 1:`
- For this, I created a script called 'Carrier stats timer'. The only content of the script is a single line to try to make it work as a timer?
`{SetState('eddi_context_approach_time', SecondsSince(0))}`
- Then, back to the main 'Carrier Stat's script above, at the end it calls the timer script with the line: 
`{F("Carrier stats timer")}`
- The idea being that once the carrier stats script is run, it calls the timer, which starts a timer, then is meant to stop the carrier stats script but is not working.

### Second attempt was: 
`{if SecondsSince(eddi_context_approach_time) > 1000:`
- In the hope that this would read the approach time. Then the line at the end of the script would start the timer:
`{SetState('eddi_context_approach_time', SecondsSince(0))}`
- But this approach isn't working either :(. 

Hopefully that explains the situation clearly? If you can see where I'm going wrong or if neither of my ideas are the best method, I look forward to hearing back o7. 

Many thanks

**Comment by Tkael** [comment]
From https://www.unixtimestamp.com/:
```
The unix time stamp is a way to track time as a running total of seconds. This count starts at the Unix Epoch on January 1st, 1970 at UTC. Therefore, the unix time stamp is merely the number of seconds between a particular date and the Unix Epoch. 
```

Thus: `SecondsSince(0)` measures the total number of seconds between the Unix epoch on Jan. 1st, 1970 and the time at which it is invoked. As you can see from the website, this is a very large number of seconds.

Your second attempt, taking the difference between the invocations of `SecondsSince()`, can be used to measure the time between events. While I believe that this approach should work, at 1000 seconds you'll need to wait at least 16.67 minutes  (1000 seconds / 60 seconds per minute) before the condition would be true.

o7

**Comment by Tkael** [comment, marked as answer]
An alternative approach might be to use `SetState()` to track the status of various `Carrier stats` variables an suppress repeats where those variables are unchanged.

**Reply by CMDR-Evolution** [reply]
> From https://www.unixtimestamp.com/:
> 
> ```
> The unix time stamp is a way to track time as a running total of seconds. This count starts at the Unix Epoch on January 1st, 1970 at UTC. Therefore, the unix time stamp is merely the number of seconds between a particular date and the Unix Epoch. 
> ```
> 
> Thus: `SecondsSince(0)` measures the total number of seconds between the Unix epoch on Jan. 1st, 1970 and the time at which it is invoked. As you can see from the website, this is a very large number of seconds.
> 
> Your second attempt, taking the difference between the invocations of `SecondsSince()`, can be used to measure the time between events. While I believe that this approach should work, at 1000 seconds you'll need to wait at least 16.67 minutes (1000 seconds / 60 seconds per minute) before the condition would be true.
> 
> o7

Thanks very much and happy new year Tkael!

Sorry about the timings seeming strange. I had it previously set to 1000 to have a play around, then would switch from 1 second to 1000 during testing. 

> An alternative approach might be to use `SetState()` to track the status of various `Carrier stats` variables an suppress repeats where those variables are unchanged.

I've managed to get it working now, thanks to your suggestions! After having another look today, I realised that I wasn't setting the state at the top of the script. 

I've removed the idea of calling a secondary script and tidied it up a bit. Here it is, up and running o7. 

```
{if SecondsSince(state.carrier_stats) > 1800:
	{set carrier to OneOf("carrier","fleet carrier")}
	{set information to OneOf("information","command","overview")}

	{if information = "overview": {set menu to OneOf("menu","display","screen")}
	|else: {set menu to OneOf("menu","view","display","screen")}}
	{set notorious to OneOf("wanted","criminal","notorious")}

	{if event.notoriousAccess: {set notoriousaccess to cat(", including ", notorious," commanders")}
	|else: {set notoriousaccess to ""}}

	{if event.dockingAccess = "all": {set dockingaccess to cat("all people", notoriousaccess)}
	|elif event.dockingAccess = "squadronfriends {notoriousaccess}": {set dockingaccess to "squadron members and friendly commanders {notoriousaccess}"}
	|elif event.dockingAccess = "friends {notoriousaccess}": {set dockingaccess to "friendly commanders only {notoriousaccess}"}
	|else: {set dockingaccess to "no commanders except yourself"}}
	{set of to OneOf("of","at")}
	
	{if event.fuel = 1000: {set tritium to "100" }
	|else: {set tritiumcalc to div(event.fuel, 1000)}
		{set tritium to mul(tritiumcalc, 100)}}

	{set cargocalc to div(event.freeCapacity, 25000)}
	{set cargo to mul(cargocalc, 100)}
	{set cargo_free to sub(100, cargo)}

	{set bankcalc to div(event.bankReservedBalance, event.bankBalance)}
	{set bank to mul(bankcalc, 100)}
	{set bank_free to sub(100, bank)}
	

	Welcome to the {OneOf(
		"{carrier}, ",
		"{carrier} {information} {SpeechRate(\"for\",\"+20%\")}, ",
		"{carrier} {information} {menu} {SpeechRate(\"for\",\"+20%\")}, ")} 
	{event.name}.

	{OneOf("The","This")} {OneOf("carrier","fleet carrier","carrier currently","The fleet carrier currently")} allows access to {dockingaccess}.

	Tritium levels are at {round(tritium)} percent or {event.fuel} tonnes.

	{OneOf("The carrier","The fleet carrier","{event.name}")}'s {OneOf("available","un-used","remaining","un-allocated")} cargo {OneOf("capacity","space","room")} is {round(cargo)} percent at {event.freeCapacity} free tonnes, 
	with {round(cargo_free)} percent or {event.usedCapacity} tonnes {OneOf("used","un-available","already allocated")}.

	The total bank balance is {Humanise(event.bankBalance)} credits.
	The balance has a {round(bank)} percent reserve {of} {Humanise(event.bankReservedBalance)}. 
	This leaves an available balance {of} {round(bank_free)} percent or {Humanise(event.bankAvailableBalance)} credits
	{OneOf("","for the carrier","for the fleet carrier","for {event.name}")}.
}

{SetState('carrier_stats', SecondsSince(0))}
```

**Comment by ChloeTrappedRabbit** [comment]
Unsubscribe

**Comment by Transcan** [comment]
I had a similar problem with "Body scanned" script: I'm too fast scanning and when I was jumping to the next system, EDDI was still telling me a body was scanned...

My approach is similar to yours. In my case the waiting time is the time that takes to speaks the actual message. So if EDDI is already talking and I scan a body meanwhile, that scan will not be reported (unless is a valuable body, in that case it will be queued).

For this method I use 2 state variables:
_state.eddi_context_body_lastscan_timestamp_ to save the scan time.
_state.eddi_context_body_lastscan_cooldown_ for saving the waiting time where no more message would be spoken.

The code goes like this:
```
{if !state.eddi_context_body_lastscan_cooldown: {SetState('eddi_context_body_lastscan_cooldown', 0)}}
...
{if ((SecondsSince(state.eddi_context_body_lastscan_timestamp) > state.eddi_context_body_lastscan_cooldown) && (system.totalbodies != system.scannedbodies)) || isWorthy(reportbody):
   {SetState('eddi_context_body_lastscan_timestamp', SecondsSince(0))}
   ...
   {SetState('eddi_context_body_lastscan_cooldown', round(len(report) * 0.05))}
   ...
}
```

That _round(len(report) * 0.05))_ is how I calculate the waiting time: 0,05 seconds per letter.

I guess in your case, the cooldown should be fixed.
But the good thing about this method is you can extend the cooldown whenever is needed: For example, increasing some more delay each time the menus are reloaded.

**Reply by CMDR-Evolution** [reply]
Hello Transcan and thanks very much for your message.

Yes, my fleet carrier script is working well now but thanks for your suggestion about the delay timer you use in your body scanned script.

I can definitely use that in other places and tried putting it into my 'Surface Signals Detected' script but don't seem to be getting any joy. In case the issue jumps out at you, would you mind having a quick look to see where I'm going wrong please?

Here's the full script:

```
{if !state.eddi_context_surfacesignals_cooldown: {SetState('eddi_context_surfacesignals_cooldown', 0)}}

{if (SecondsSince(state.eddi_context_surfacesignals_timestamp) > state.eddi_context_surfacesignals_cooldown):
	{SetState('eddi_context_surfacesignals_timestamp', SecondsSince(10))}
	{set surfacesignals to []}
	{if vehicle = "Ship": {set theship to OneOf("{F(\"Ships\")}","{F(\"Ships\")}","{F(\"Ships\")}","{F(\"Ships\")}","The ship")}
	|else: {set theship to "The ship"}}

	{set theshipis to OneOf("{theship} is","{theship} is","{theship} is","{theship} is","{theship}'s","{theship}'s","")}
	{set sensors to OneOf(
			"sensors","scanners","sensor results","scanner results","sensor systems","scanner systems",
			"sensor system results","scanner system results","sensor module results","scanner module results")}

	{if theship = "The ship": {set thesensors to: {Occasionally(2,"the")} {sensors}}
	|else: {set thesensors to: {Occasionally(2, OneOf("the","the ships"))} {sensors}}}

	{set reads to OneOf("reads","displays")}
	{set detecting to OneOf("picking up","detecting","reading","showing","finding")}
	{set detected to OneOf("picked up","detected","shown","recorded","confirmed","found")}
	{set pick_up to OneOf("pick up","detect","read","show","confirm","find")}

	{for signal in event.surfacesignals: 
	    {set surfacesignals_desc to: 
	        {signal.amount} {signal.source}{if signal.amount > 1:s}
	    }
	    {set surfacesignals to cat(surfacesignals, [surfacesignals_desc])}
	}

	{set body to BodyDetails(event.bodyname, system.systemname)}

	{if List(surfacesignals) > 1: {set are to OneOf("are","")}
	|else: {set are to OneOf("is","")}}

	{if body:
		{set report to: {OneOf(
			"{theshipis} {detecting} {List(surfacesignals)} {OneOf(\"on {thesensors}, for\",\"for\")} {body.bodytype} {P(body.shortname)}",
			"{theship}'s {OneOf(\"sensors\",\"scanners\",\"sensor results\",\"scanner results\",\"sensor systems\",\"scanner systems\",\"sensor system results\",\"scanner system results\",\"sensor module results\",\"scanner module results\")}
				for {body.bodytype} {P(body.shortname)} {OneOf(\"{pick_up}\",\"are {detecting}\")} {List(surfacesignals)}",
			"{theship} {reads} {body.bodytype} {P(body.shortname)} {OneOf(\"as having\",\"with\",\"has\")} {List(surfacesignals)}",
			"{List(surfacesignals)} {are} {detected} on {thesensors} for {body.bodytype} {P(body.shortname)}",
			"{body.bodytype} {P(body.shortname)} is showing {List(surfacesignals)}{OneOf(\"on {thesensors}\",\"\")}",
			"{body.bodytype} {P(body.shortname)} {reads} {List(surfacesignals)} on {thesensors}",
			"{thesensors} {OneOf(\"{pick_up}\",\"are {detecting}\")} {List(surfacesignals)} for {body.bodytype} {P(body.shortname)}")}.
		}
	|else:
		{set report to: {OneOf(
			"{theshipis} {detecting} {List(surfacesignals)} on {thesensors}",
			"{theship}'s {OneOf(\"sensors\",\"scanners\",\"sensor results\",\"scanner results\",\"sensor systems\",\"scanner systems\",\"sensor system results\",\"scanner system results\",\"sensor module results\",\"scanner module results\")} 
				{OneOf(\"{pick_up}\",\"are {detecting}\")} {List(surfacesignals)}",
			"{theship} {reads} {List(surfacesignals)}",
			"{List(surfacesignals)} {are} {detected} on {thesensors}",
			"{detecting} {List(surfacesignals)} on {thesensors}",
			"{thesensors} {OneOf(\"{pick_up}\",\"are {detecting}\")} {List(surfacesignals)}")}.
		}
}
	{report}
	{SetState('eddi_context_surfacesignals_timestamp', round(len(report) * 0.05))}
}
```

**Reply by Transcan** [reply]
Yes, I tried to apply that method to that script with no result.
After some investigation I noticed the script is triggered for every signal detected at once.
I mean, all the signals are triggered at the same time with several threads of the same script and the messages are queued one after another.
So this method doesn't work (and I don't know how to handle in another way).
Maybe Devs can enlighten us with another way to achieve it.

**Reply by CMDR-Evolution** [reply]
Oh cool, glad it wasn't just me. Thanks for having a look and thanks again for the info about delay timers o7

**Comment by Welgum** [comment]
Unsubscribe
